### ISSUE TITLE
UUID/Name not updated with new name after name change

### ISSUE THREAD

**Issue by AeSix** [IntellectualSites/PlotSquared#2450]
__*NOTICE: Bukkit/Spigot versions 1.7.10 to 1.12.2 are considered legacy and will receive limited support. Please consider upgrading to 1.13 for future support. Plugins exist for 1.13+ that bring back old behaviors found in 1.8*__
# Bug report template 

**[REQUIRED] PlotSquared Version Number:** 

**[REQUIRED] Spigot/Paper Version Number:** 
1.13.2 / git-Spigot-1a3504a-84f3da3  (final build)
**[REQUIRED] Minecraft Version Number:** 
1.13.2
**Links to worlds.yml file and settings.yml file:** 
https://athion.net/ISPaster/paste/view/362c661dc1da47cfb19b10f8e11a9578

**[REQUIRED] Description of the problem:** 
I changed my IGN from "AeSix_Reficul" to "AeSix" 
Upon joining my P2 server, my plots are all identified as having an owner of "____Unknown____"
Certain commands and functions treat the plots as not having an owner. Most in fact, include use flags, /p trust, and others. 
Attempting to /p setowner states I am already the owner of the plot.
HOWEBER, I am able to use /p home to go to any of my plots. 

Additionally, I stopped the server, removed usercache, and started the server, with no change in this issue.

**Any relevant console output or screenshots:**
no errors in console. No output of any kind from these commands, or relevancy to this issue.

**Plugins being used on the server:**
WorldEdit, OpenInv, CreativeNbtControl, CommandHelper, LogBlock, Craftconomy3, ScheduledAnnouncer2, EasySetSpawn, STFU, WorldBorder, ProtocolLib, RemoveGodItems, LibsDisguises, RedstoneClockDetector, LogBlockQuestioner, ModReq, PermissionsEx, Vault, CommandBook, StopPushing, ColoredAnvils,  HolographicDisplays, PlotSquared, QuickShop, BuycraftX, WorldGuard, OnTime, VanishNoPacket, LockettePro, EditableSign, RandomTeleport, PurpleIRC, 

I have a feeling this may be something to do with name vs UUID somewhere. 

**How to replicate:** 

Not practical to replicate unless someone is willing to change their account name (IGN). 
1) join a p2 server, claim plots and log out
2) change IGN 
3) re-join the server
4) attempt commands, view plot info, signs, holograms
**Checklist**:

- [X] I included all information required in the sections above
- [X] I made sure there are no duplicates of this report [(Use Search)](https://github.com/IntellectualSites/PlotSquared/issues?utf8=%E2%9C%93&q=is%3Aissue)
- [X] I made sure I am using an up-to-date version of PlotSquared
- [X] I made sure the bug/error is not caused by any other plugin

**Comment by MattBDev** [comment]
@AeSix please mark as high priority. (I'm on mobile so I can't)

**Comment by AeSix** [comment]
Done and added testing required, hoping someone else will be able to confirm this.

**Comment by zachrocks22** [comment]
Can confirm on 1.14.4 and latest PlotSquared version.

**Comment by AeSix** [comment]
YAY! I wasn't losing my mind - just my identity >_<

**Comment by AeSix** [comment]
I will be able to test after August 28th.  I have another account which I will be changing the name on.  I will get with Matt before testing this to ensure that it's actually ready to be tested.

**Comment by AeSix** [comment]
@MattBDev 
I just tested with P2 4.340, went on and was able to go to my home, etc. 
I changed my alt from "ArchBeingSix" to "AeSix_Reficul"  - The /p info and signs both still read "ArchBeingSix" - which is a LOT better than "Unknown" 

Perhaps this is a UUID caching issue at this point?

**Comment by MattBDev** [comment]
@Aesix I'm going to fix the delay for Mojang API requests. The thread will sleep for 10 minutes and then resume. I am thinking about adding a message for the user on login to let them know it may take awhile for their name to update. I'll update necessary commands to make sure they are all relying on UUID's instead of names when verifying the owner.

**Comment by AeSix** [comment]
Sounds great!  We'll need someone to test though.  I have a staff who likes to change their name... I'll ask if they'll be doing that soon.

**Comment by RedstoneFuture** [comment]
Maybe the same problem: https://github.com/IntellectualSites/PlotSquared/issues/2467

**Comment by AeSix** [comment]
I didn't expeirence the more plots issue - but that's not to say it's not a symptom of the same issue.

**Comment by boy0001** [comment]
note: this shouldn't need mojang api requests, since as mentioned, the player joined the server and it was still "Unknown". The server should have the UUID & name already.

**Comment by zachrocks22** [comment]
I agree with boy0001. I feel like asking Mojang would be ultimately unnecessary and resource intensive. If utilizing Mojang's online API for UUIDs is absolutely necessary, it should be done only on startup if PlotSquared is having extreme issues finding the players UUID as a last resort.

**Comment by AeSix** [comment]
You should NEVER batch lookup UUIDs on startup.  This is stupid and completely useless.  It ties up the entire server while doing the lookups - and if 200k players have joined that server, it's going to do 200k lookups.  That server will, well before it completes lookups, be blacklisted from the API servers.  

The UUID API is to be used on player join, when that player's cached info either does not exist, or has expired.  For both, the plugin and the server's cache.  

Anyone doing full list lookups on startup should be taken behind the barn and shot like a lame horse.!

**Comment by zachrocks22** [comment]
“...only... if PlotSquared is having extreme issues finding the players UUID.“

You wouldn’t look up everyone’s UUID, you could queue lookups for “””extreme issues””” where P2 is struggling to find a reliable String for the UUID.  Which would hopefully only be a player or two.

As for holding up the main server thread, you could run these checks Asynchronously.

**Comment by AeSix** [comment]
There's two reasons to do a player UUID lookup - from plugin or server cache or Mojangs API: On player join, and on player info/manipulation (assigning perms, (un)banning, plot ownerships, etc) - If the player's UUID cannot be found in the plugin's cache (if it has one), or the server's cache, then that is when the UUID should be queried through the API.  Again, API server lookups should happen only once in a while for each player, and at join or on cache expiration/non-existence.

If the server cannot get valid information from the API servers, then there's either a connectivity issue (failed routes, blacklisted), the API server is down, or some plugin (or the mc server) is screwing things up.   The proper action is to find the actual problem and correct it. 

Since we're talking about PlotSquared specifically, this is what one would assume is the proper list of actions:

Player UUID is needed (for 1 of many reasons):
- PS looks at it's own cache, and (finds a match - done | does not find it - proceeds)
- PS looks at the server's cache (finds its, stores it - done | does not find it - proceeds)
- PS/server queries Mojang's API server (finds it, caches it, done | does not find it - fails)
- On failure, there is either a connectivity issue, a malformed lookup, or the API server does not have that username.  

PS needs a player name /from/ a UUID (again, for 1 of many reasons):
- PS looks in it's cache (finds a valid entry, done | finds an expired entry - proceeds | finds the UUID, but with a different name - proceeds | does not find it - proceeds)
- PS looks in the server cache (finds a valid entry - done | finds expired entry - proceeds | finds the UUID, but with a different name - proceeds | does not find it - proceeds)
- API server is queried (finds the UUID/name - done | does not find a valid UUID - failure)
- On failure, the UUID is not authentic (or Mojang's API server is having issues, or the query was malformed/wrong)
At this point the only thing that can be done is to re-query the API server at a later time in hopes of finding a match.  If after a few attempts, the UUID should be marked as invalid and no additional lookups performed.  

And obviously, no offline UUIDs lookups should ever make it past the server cache, as those are neither authentic. 

Other than the UUID or player name not being authentic (held within Mojang's API), connectivity issues, or malformed requests - there is no reason a lookup should fail plugin and server cache AND API server.  Brute-forcing it does nothing but waste cycles and storage space for the additional code. 

What's NOT happening, is PS is not catching the name change, not registering the name change, or is waiting on cache expiry before doing anything with the name change.  With there being failed reproduction results, I would err on the side of the latter, with timing being the main issue.  Which can be fixed. 

Now, there may be some other niche reasons why lookups could fail, and there may be reasons why plugin/server cache lookups could fail for offline UUIDs not provided here - but those are already addressed for the most part.

**Comment by N0tMyFaultOG** [comment]
No response, assume fixed.
